doc,bgpv1: Bootstrapping BGP CPlane failure scenario doc (#31153)
cilium-team/cilium: Documentation/network/bgp-control-plane-failure-scenarios.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _bgp_control_plane_failure_scenarios:

> BGP Control Plane Failure Scenarios

Reviewer: To me this feels like it should be part of the existing "BGP Control Plane Troubleshooting Guide", not a separate document. As a user, I would probably have to look into both to find out if one of them contains a solution to my problem.

What I would suggest is to extend the existing "BGP Control Plane Troubleshooting Guide" and have two sections in it, e.g.:
 - Configuration Issues (the existing content would go under this)
 - Operational Issues (the content of this PR would go there)

Author: The timing of reading is different for those two. "Trouble Shooting Guide" is a document to see when trouble occurs, but for this kind of failure scenario, it is too late to see it when it occurs. The user should be prepared instead.

If it is a part of "Operational Guide" for example, that's understandable, but for "Trouble Shooting" that's questionable for me.

I don't have a super strong opinion, though. @networkop, what do you think? You 👍 to the above comment, so maybe you agree with Rastislav here?

Reviewer 2: Yeah, I +1'ed before I read the entire document. Now it does feel like it's less of a troubleshooting guide and more of a "Failover scenarios" inside of the operational guide.

Reviewer: ACK, makes sense to me as well now, thanks for the explanation

Author: Create a new Operation Guide doc and move the contents into there.

===================================

This document describes common failure scenarios that you may encounter when
using the BGP Control Plane and provides guidance on how to mitigate them.

Cilium Agent Down
-----------------

If the Cilium agent goes down, the BGP session will be lost because the BGP
speaker is integrated within the Cilium agent. The BGP session will be restored
once the Cilium agent is restarted. However, while the Cilium agent is down,
the advertised routes will be removed from the BGP peer. As a result, you may
temporarily lose connectivity to the Pods or Services.

Mitigation
~~~~~~~~~~

The recommended way to address this issue is by enabling the
:ref:`bgp_control_plane_graceful_restart` feature. This feature allows the BGP
peer to retain routes for a specific period of time after the BGP session is
lost. Since the datapath remains active even when the agent is down, this will
prevent the loss of connectivity to the Pods or Services.

> When you can't use BGP Graceful Restart, you can take following actions depending
> on what kind of the route you are using:

Reviewer 3: ```suggestion
When you can't use BGP Graceful Restart, you can take the following actions, depending
on the kind of routes you are using:
```

> .. _BGP Graceful Restart: 

Reviewer 3: What is this label for? I can't see it referenced anywhere

Author: Thanks, deleted.

> **PodCIDR routes**

Reviewer 3: Nit: Use proper (lower-level) heading?

If you are advertising PodCIDR routes, pods on the failed node will be
unreachable from the external network. If the failure only occurs on a subset

> of the nodes in the cluster, you can drain the unhealthy node to migrate the

Reviewer 3: ```suggestion
of the nodes in the cluster, you can drain the unhealthy nodes to migrate the
```

“Subset of nodes” means potentially several nodes

Service routes
++++++++++++++

If you are advertising service routes, the load balancer (KubeProxy or Cilium
KubeProxyReplacement) may become unreachable from the external network.
Additionally, ongoing connections may be redirected to different nodes due to
ECMP rehashing on the upstream routers. When the load balancer encounters
unknown traffic, it will select a new endpoint. Depending on the load
balancer's backend selection algorithm, the traffic may be directed to a
different endpoint than before, potentially causing the connection to be reset.

> If your upstream routers support ECMP with consistent hashing, enabling it may

Reviewer 2: I believe in the hardware world the more common term is  "resilient" hashing.

help to keep the ongoing connections forwarded to the same node. Enabling the
:ref:`maglev` feature in Cilium may also help since it increases
the probability that all nodes select the same endpoint for the same flow.

> However, it only works for the "externalTrafficPolicy: Cluster". If the
> Service's "externalTrafficPolicy" is set to "Local", it is inevitable that all

Reviewer 3: ```suggestion
However, it only works for the ``externalTrafficPolicy: Cluster`` policy. If the
Service's ``externalTrafficPolicy`` is set to ``Local``, it is inevitable that all
```

Author: Oops, seems like grammar collector modified that... Fixed.

ongoing connections with the endpoints on the failed node and connections

> forwarded to a different node than before will be reset.

Reviewer 3: Long sentence, difficult to parse because the verb is far away from “it is inevitable that...”. I'd maybe add a few commas.

```suggestion
ongoing connections with the endpoints on the failed node, and connections
forwarded to a different node than before, will be reset.
```

Alternatively, “a reset will be inevitable for all ongoing connections...”.
